### PR TITLE
Random order generation

### LINKED ISSUE #25: Write a function that generates random DAIDE orders
Given a set of DAIDE orders, create a set of random orders that is similar.

orders might be "unit HLD (prov), unit SUP unit..."

Use this DAIDE parser linked here as needed: https://github.com/ALLAN-DIP/baseline_bots/issues/16

```python

def random_orders(orders:str):

   return some random orders...

### PR DESCRIPTION
**Pull request by kkahadze**
Fixes #25

### CODE REVIEW

File: src/baseline_bots/utils.py [hunk 1]
```diff
@@ -15,6 +17,229 @@
...
+COMBOS = {
+    "FLT" : {"FLT": {"WATER", "COAST"}, "AMY": {"COAST"}},
+    "AMY" : {"FLT": {"COAST"}, "AMY": {"LAND", "COAST"}}
+}
+# fmt: on
+
+
+def random_orders(orders: list):
```
Comment (trigaten) [inline src/baseline_bots/utils.py:30]: `list` should be `List`

File: src/baseline_bots/utils.py [hunk 2]
```diff
@@ -15,6 +17,229 @@
 from diplomacy import Game, Message
 from tornado import gen
 
+# fmt: off
```
Comment (trigaten) [inline src/baseline_bots/utils.py:20]: Can you provide some minimal comments which explain these constants.

File: src/baseline_bots/utils.py [hunk 3]
```diff
@@ -15,6 +17,229 @@
...
+
+        new = list(map(lambda order: randomize(order), new))
+        new.extend(replacements)
+        return new
+    else:
+        new = list(map(lambda order: randomize(order), orders))
+        return new
+
+
+def orders_correspondence(orders: list) -> (list):
```
Comment (trigaten) [inline src/baseline_bots/utils.py:69]: What is (list)?

**Review by trigaten — CHANGES_REQUESTED** [review]
Great docstrings!

A bit of work to be done on type hints. Also, I would like to see more regular (#) comments throughout the code.

Finally, please move your functions into a new file.

File: src/baseline_bots/utils.py [hunk 4]
```diff
@@ -15,6 +17,229 @@
...
+            for adjacency in sup_adjacent
+            if adjacency in rec_adjacent and adjacency != province and TYPES[adjacency]
+        ]
+        if adj_to_both:
+            return (order[0], tag, order[2], "MTO", random.choice(adj_to_both))
+        else:
+            return order
+
+
+def random_movement(order, chance_of_move=0.3):
```
Comment (trigaten) [inline src/baseline_bots/utils.py:212]: type hint order

File: tests/utils_test.py [hunk 5]
```diff
@@ -41,4 +41,22 @@ def test(self):
         game.add_message(message=msg_obj2)
         msgs = [msg_obj2, msg_obj1]
 
-        assert sort_messages_by_most_recent(msgs)[0].message == "HELLO"
\ No newline at end of file
+        assert sort_messages_by_most_recent(msgs)[0].message == "HELLO"
+
+        ord =  [(('FRA', "AMY", 'PAR'), 'BLD')]
```
Comment (trigaten) [inline tests/utils_test.py:46]: provide comments that explain the test cases

File: src/baseline_bots/randomize_dipnet.py [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+an already existing order / list of orders.
+"""
+
+__author__ = "Konstantine Kahadze"
+__email__ = "konstantinekahadze@gmail.com"
+
+import random
+from typing import List
+
+# the comments below signal the formatter not to expand these dicts to multiple lines
```
Comment (trigaten) [inline src/baseline_bots/randomize_dipnet.py:12]: Sorry, I meant comments to describe ADJACENCY, TYPES, and COMBOS. What is the point of each of these constants?

File: src/baseline_bots/randomize_dipnet.py [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+    "FLT" : {"FLT": {"WATER", "COAST"}, "AMY": {"COAST"}},
+    "AMY" : {"FLT": {"COAST"}, "AMY": {"LAND", "COAST"}}
+}
+
+joiners = {'AND', 'ORR'}
+
+# fmt: on
+
+
+def random_orders(orders: List):
```
Comment (trigaten) [inline src/baseline_bots/randomize_dipnet.py:26]: Need return type hint

File: src/baseline_bots/randomize_dipnet.py [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+    Takes in a list of orders in the following form:
+
+    [(("FRA", "AMY", "PIC"), "MTO", "PAR"), (("FRA", "AMY", "BUR"), "SUP", ('FRA', 'AMY', "PIC"),"MTO", "PAR")]
+
+    and returns randomly deviant orders in the same form.
+    """
+    correspondences = orders_correspondence(orders)
+
+    if correspondences:
+        new = orders
```
Comment (trigaten) [inline src/baseline_bots/randomize_dipnet.py:37]: Please rename the variable `new` to be more descriptive.

File: src/baseline_bots/randomize_dipnet.py [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+        "HLD": random_hold,
+        "SUP": random_support,
+        "CVY": random_convoy,
+        "CTO": random_convoy_to,
+    }
+
+    return tag_to_func[tag](order)
+
+
+def random_convoy_to(order: tuple) -> tuple:
```
Comment (trigaten) [inline src/baseline_bots/randomize_dipnet.py:110]: `tuple` should be `Tuple`. Import the `Tuple` type.

File: src/baseline_bots/randomize_dipnet.py [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+    """
+    tag = order[1]
+    (
+        (amy_country, amy_type, amy_loc),
+        _,
+        (flt_country, flt_type, flt_loc),
+        _,
+        province,
+    ) = order
+    assert amy_type == "AMY" and flt_type == "FLT"
```
Comment (trigaten) [inline src/baseline_bots/randomize_dipnet.py:145]: Make this return an intelligent error string if the assertion fails

File: src/baseline_bots/randomize_dipnet.py [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+        for loc in ADJACENCY[flt_loc]
+        if TYPES[loc] == "COAST" and loc not in ADJACENCY[amy_loc] and loc != province
+    ]
+    if adj:
+        return (order[0], tag, order[2], "CTO", random.choice(adj))
+    else:
+        return order
+
+
+def random_support(order: tuple) -> tuple:
```
Comment (trigaten) [inline src/baseline_bots/randomize_dipnet.py:157]: Again, correct `tuple`

File: src/baseline_bots/randomize_dipnet.py [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+    """
+    Takes in a support order and returns a believable but randomized version of it.
+    """
+    tag = order[1]
+    if len(order) <= 3:  # if it is supporting to hold
```
Comment (trigaten) [inline src/baseline_bots/randomize_dipnet.py:162]: # comments throughout would be good

File: tests/randomize_dipnet_test.py [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+        assert ord == random_orders(ord)
+
+        ord =  [(('FRA', "AMY", 'PAR'), 'DSB')]
+        assert ord == random_orders(ord)
+
+        ord =  [(('FRA', "AMY", 'PAR'), 'REM')]
+        assert ord == random_orders(ord)
+
+        # The following three tests check that when orders that contain movements, holds, convoys and other moves get
+        # input into the order randomizer, they come out differnent
```
Comment (trigaten) [inline tests/randomize_dipnet_test.py:19]: I would like to see a seeded trial, where you compare the output to an actual expected string

Comment (trigaten) [reply tests/randomize_dipnet_test.py:19]: also different*

**Review by trigaten — CHANGES_REQUESTED** [review]

**Comment by trigaten** [comment]
Please resolve merge conflicts and add a sphinx page for your functions.

File: src/baseline_bots/randomize_dipnet.py [hunk 14]
```diff
@@ -0,0 +1,320 @@
...
+                    correspondences.append((order, (u2, "HLD")))
+        elif order[1] == "CVY":
+            convoy_moves = filter(  # filter all the convoy related moves
+                lambda order: order[1] == "CTO" or order[1] == "CVY", orders
+            )
+            correspondences.append(tuple(convoy_moves))
+    return correspondences
+
+
+def randomize(order: tuple) -> tuple:
```
Comment (trigaten) [inline src/baseline_bots/randomize_dipnet.py:106]: Fix all of the tuple -> Tuple problems

File: src/baseline_bots/randomize_dipnet.py [hunk 15]
```diff
@@ -0,0 +1,320 @@
...
+        ((country, unit_type, loc), _) = order
+        move_loc = random.choice(
+            ADJACENCY[loc]
+        )  # randomly chooses an adjacent location
+        return ((country, unit_type, loc), "MTO", move_loc)
+    else:
+        return order
+
+
+def random_build(order: Tuple) -> Tuple:
```
Comment (trigaten) [inline src/baseline_bots/randomize_dipnet.py:258]: What?

File: src/baseline_bots/randomize_dipnet.py [hunk 16]
```diff
@@ -0,0 +1,320 @@
...
+    with_commas = re.sub(
+        r"(.*?[^(])\s+?([^)].*?)", r"\1, \2", orders
+    )  # inserts commas in between tuples and strings
+    with_quotes = re.sub(
+        r"([(, ])([A-Z]+)([), ])", r"\1'\2'\3", with_commas
+    )  # inserts quotes around strings
+    return eval(with_quotes)
+
+
+def randomize_joiner(order: str) -> str:
```
Comment (trigaten) [inline src/baseline_bots/randomize_dipnet.py:296]: Can you rename this function to smtg more intelligible like `randomize_order`?

File: src/baseline_bots/randomize_dipnet.py (hunk 15 above)
Comment (kkahadze) [reply src/baseline_bots/randomize_dipnet.py]: I fixed that

**Comment by trigaten** [comment]
Can you rename the file name to randomize_order

File: tests/randomize_dipnet_test.py [hunk 17]
```diff
@@ -0,0 +1,53 @@
+from baseline_bots.randomize_dipnet import random_orders, string_to_tuple, tuple_to_string, randomize_joiner
+from typing import Tuple
+import random
+class TestUtils:
```
Comment (trigaten) [inline tests/randomize_dipnet_test.py:4]: rename this appropriately

**Review by trigaten — APPROVED** [review]
Great job Konstantine! The code looks wonderful. 

Take a look at the recommendations here: https://allan-dip.github.io/baseline_bots/development/#pushing-code

**Pull request merged**